crypto/threads_pthread.c

```diff
@@ -42,12 +44,481 @@
+        /* we have to wait for one to be free */
+        pthread_cond_wait(&ilock->alloc_signal, &ilock->alloc_lock);
+    }
+
+    current_idx = ilock->current_alloc_idx;
+
+    /* Allocate the qp */
+    ilock->writers_alloced++;
+
+    /* increment the allocation index */
+    ilock->current_alloc_idx =
+        (ilock->current_alloc_idx + 1) % ilock->group_count;
+
+    /* get and insert a new id */
+    new_id = __ATOMIC_FETCH_ADD(&ilock->id_ctr, 1, __ATOMIC_SEQ_CST);
+
+    new_id = VAL_ID(new_id);
+    __ATOMIC_AND_FETCH(&ilock->qp_group[current_idx].users, ID_MASK,
+                       __ATOMIC_SEQ_CST);
+    __ATOMIC_OR_FETCH(&ilock->qp_group[current_idx].users, new_id,
+                      __ATOMIC_SEQ_CST);
+
+    /* update the reader index to be the prior qp */
+    __ATOMIC_STORE_N(&ilock->reader_idx, ilock->current_alloc_idx,
+                     __ATOMIC_SEQ_CST);
+
+    /* wake up any waiters */
+    pthread_cond_signal(&ilock->alloc_signal);
+    pthread_mutex_unlock(&ilock->alloc_lock);
+    return (struct rcu_qp *)&ilock->qp_group[current_idx];
+}
+
+static void retire_qp(struct rcu_lock_internal *ilock,
+                      struct rcu_qp *qp)
+{
+    pthread_mutex_lock(&ilock->alloc_lock);
+    ilock->writers_alloced--;
+    pthread_cond_signal(&ilock->alloc_signal);
+    pthread_mutex_unlock(&ilock->alloc_lock);
+}
+
+static struct rcu_qp *allocate_new_qp_group(struct rcu_lock_internal *ilock,
+                                            int count)
+{
+    struct rcu_qp *new =
+        OPENSSL_zalloc(sizeof(struct rcu_qp) * count);
+
+    ilock->group_count = count;
+    return new;
+}
+
+/* Public WRITE api */
+void ossl_rcu_write_lock(CRYPTO_RCU_LOCK *lock)
+{
+    struct rcu_lock_internal *ilock = (struct rcu_lock_internal *)lock;
+
+    pthread_mutex_lock(&ilock->write_lock);
+}
+
+void ossl_rcu_write_unlock(CRYPTO_RCU_LOCK *lock)
+{
+    struct rcu_lock_internal *ilock = (struct rcu_lock_internal *)lock;
+
+    pthread_mutex_unlock(&ilock->write_lock);
+}
+
+/* Public WRITE api for synchronization */
+void ossl_synchronize_rcu(CRYPTO_RCU_LOCK *lock)
+{
+    struct rcu_qp *qp;
+    struct rcu_lock_internal *ilock;
+    uint64_t count;
+    struct rcu_cb_item *cb_items, *tmpcb;
+
+    /* before we do anything else, lets grab the cb list */
+    ilock = (struct rcu_lock_internal *)lock;
+    cb_items = __ATOMIC_EXCHANGE_N(&ilock->cb_items, NULL, __ATOMIC_SEQ_CST);
+
+    qp = update_qp(ilock);
+
+    /* wait for the reader count to reach zero */
+    do {
+        count = (uint64_t)__ATOMIC_LOAD_N(&qp->users, __ATOMIC_SEQ_CST);
+    } while (READER_COUNT(count) != 0);
+
+    /* retire in order */
+    pthread_mutex_lock(&ilock->prior_lock);
+    while (ilock->next_to_retire != ID_VAL(count))
+        pthread_cond_wait(&ilock->prior_signal, &ilock->prior_lock);
+    ilock->next_to_retire++;
+    pthread_cond_broadcast(&ilock->prior_signal);
+    pthread_mutex_unlock(&ilock->prior_lock);
```

Comment:
writers_alloced can never be > 1 ....so is all this necessary?

Reply:
It can If the qp group count is ever increased above the value of two, and I have infrastructure available to make that possible in the future, so I would rather keep this in place, given that its largely a no-op in the group count == 2 case

Reply:
I could expand the api now to allow lock creators to specifify a number of write side waiters if you like

Reply:
Well...I guess it depends whether we think we are going to use that capability or not. There seems little point to me in having additional complexity for no purpose - "just in case". So either we fully implement multiple writers, or we take the decision to only have 1 and remove the complexity.

Reply:
Increasing the group count gives us incrementally improved performance on the write side.  For instance upping the group count on the thread test gives us about 10% more writes per unit time.  So an improvement, but one for a lock type for which the expectation is limited writing.  i can expand the api to accomodate this and default it to 2

Reply:
I've exposed the group_count value to the rcu_lock_new api